### ISSUE TITLE
reject image doesn't go to rejected image filter anymore

### ISSUE THREAD

**Issue by Nilvus** [darktable-org/darktable#4823]
**Describe the bug**
If I reject image with "all except rejected images" filter activated on lighttable, the image should disappear when rejected but stay on lighttable view. If I change filter mode on top left to rejected images only, rejected images doesn't appear anymore.

I've tested that on my personal config but also with a default config folder with only a set of imported images: same issue.

**Expected behavior**
Apply as before rejected image in regards of activated filter mode. So having all rejected images visible on all images and rejected only ones only.

**Platform (please complete the following information):**
darktable master ; Ubuntu 20.04

@TurboGit or @AlicVB: could you confirm that? Could you be one of your latest PR/commit about reject toggle or lighttable work?

**Comment by phweyland** [comment]
On dt master / debian it works for me.
With "all except rejected images", reject an image it disappears, ctrl-z it appears again...

**Comment by Nilvus** [comment]
@phweyland: thanks for your reply. Strange, I've just a new test with last commit (3.1.0+1339~g20080a396) with all fresh new build so, no config/darktable folder and so no image. I've then launch a so totally new fresh darktable. I then imported a set of images that doesn't have any XMP. So, that exclude an issue from any personal config (database, config folder or XMP totally excluded ; I even remove my darktable cache, even if it's normally not related to something like that) and I still have this issue. This issue has appear last week, few days ago that was working.

I don't see any other thing that could be on my system to have such issue. Really strange.

**Comment by Nilvus** [comment]
maybe the debug output could help:

```
138,308680 [sql] /home/nilvus/Projets/darktable/src/common/selection.c:152, function dt_selection_clear(): exec "DELETE FROM main.selected_images"
138,308901 [sql] /home/nilvus/Projets/darktable/src/common/tags.c:623, function dt_set_darktable_tags(): prepare "SELECT COUNT(*) FROM memory.darktable_tags"
138,308954 [sql] /home/nilvus/Projets/darktable/src/common/tags.c:660, function dt_tag_get_attached(): prepare "SELECT DISTINCT I.tagid, T.name, T.flags, T.synonyms, COUNT(DISTINCT S.imgid) AS inb FROM main.selected_images AS S JOIN main.tagged_images AS I ON I.imgid = S.imgid JOIN data.tags AS T ON T.id = I.tagid WHERE T.id NOT IN memory.darktable_tags GROUP BY I.tagid ORDER by T.name"
138,309102 [sql] /home/nilvus/Projets/darktable/src/common/tags.c:1224, function dt_selected_images_count(): prepare "SELECT count(*) FROM main.selected_images"
138,311088 [sql] /home/nilvus/Projets/darktable/src/common/tags.c:623, function dt_set_darktable_tags(): prepare "SELECT COUNT(*) FROM memory.darktable_tags"
138,311135 [sql] /home/nilvus/Projets/darktable/src/common/tags.c:660, function dt_tag_get_attached(): prepare "SELECT DISTINCT I.tagid, T.name, T.flags, T.synonyms, COUNT(DISTINCT S.imgid) AS inb FROM main.selected_images AS S JOIN main.tagged_images AS I ON I.imgid = S.imgid JOIN data.tags AS T ON T.id = I.tagid WHERE T.id NOT IN memory.darktable_tags GROUP BY I.tagid ORDER by T.name"
138,311234 [sql] /home/nilvus/Projets/darktable/src/common/tags.c:1224, function dt_selected_images_count(): prepare "SELECT count(*) FROM main.selected_images"
138,312919 [sql] /home/nilvus/Projets/darktable/src/common/collection.c:914, function dt_collection_get(): prepare "SELECT id FROM main.selected_images AS s JOIN (SELECT DISTINCT mi.id FROM images AS mi WHERE   (flags & 256) != 256  AND  (flags & 7) >= 0 AND (flags & 7) != 6  AND  (1=1 AND (film_id IN (SELECT id FROM main.film_rolls WHERE folder LIKE '/home/nilvus/Images/Photos/Négatif RAW/2020/04-2020/Seyssinet'))) ORDER BY datetime_taken, group_id , mi.id-group_id != 0, filename, version LIMIT ?1, ?2) AS mi WHERE mi.id = s.imgid LIMIT -1, ?3"
138,313336 [sql] /home/nilvus/Projets/darktable/src/common/collection.c:893, function dt_collection_get_selected_count(): prepare "SELECT COUNT(*) FROM main.selected_images"
138,336351 [lighttable] expose took 0,0000 sec
138,457096 [sql] /home/nilvus/Projets/darktable/src/common/image_cache.c:239, function dt_image_cache_write_release(): prepare "UPDATE main.images SET width = ?1, height = ?2, filename = ?3, maker = ?4, model = ?5, lens = ?6, exposure = ?7, aperture = ?8, iso = ?9, focal_length = ?10, focus_distance = ?11, film_id = ?12, datetime_taken = ?13, flags = ?14, crop = ?15, orientation = ?16, raw_parameters = ?17, group_id = ?18, longitude = ?19, latitude = ?20, altitude = ?21, color_matrix = ?22, colorspace = ?23, raw_black = ?24, raw_maximum = ?25, aspect_ratio = ROUND(?26,1), exposure_bias = ?28 WHERE id = ?27"
138,457535 [sql] /home/nilvus/Projets/darktable/src/common/image.c:220, function dt_image_full_path(): prepare "SELECT folder || '/' || filename FROM main.images i, main.film_rolls f WHERE i.film_id = f.id and i.id = ?1"
138,457652 [sql] /home/nilvus/Projets/darktable/src/common/image.c:305, function dt_image_path_append_version(): prepare "SELECT version FROM main.images WHERE id = ?1"
138,457706 [sql] /home/nilvus/Projets/darktable/src/common/image.c:220, function dt_image_full_path(): prepare "SELECT folder || '/' || filename FROM main.images i, main.film_rolls f WHERE i.film_id = f.id and i.id = ?1"
138,457774 [sql] /home/nilvus/Projets/darktable/src/common/image.c:248, function _image_local_copy_full_path(): prepare "SELECT folder || '/' || filename FROM main.images i, main.film_rolls f WHERE i.film_id = f.id AND i.id = ?1"
138,463393 [sql] /home/nilvus/Projets/darktable/src/common/exif.cc:3344, function _exif_xmp_read_data(): prepare "SELECT filename, flags, raw_parameters,        longitude, latitude, altitude, history_end, datetime_taken FROM main.images WHERE id = ?1"
138,463676 [sql] /home/nilvus/Projets/darktable/src/common/iop_order.c:316, function dt_ioppr_get_iop_order_version(): prepare "SELECT version FROM main.module_order WHERE imgid = ?1"
138,463739 [sql] /home/nilvus/Projets/darktable/src/common/iop_order.c:607, function dt_ioppr_get_iop_order_list(): prepare "SELECT version, iop_list FROM main.module_order WHERE imgid=?1"
138,463867 [sql] /home/nilvus/Projets/darktable/src/common/exif.cc:3295, function dt_set_xmp_dt_metadata(): prepare "SELECT key, value FROM main.meta_data WHERE id = ?1"
138,463952 [sql] /home/nilvus/Projets/darktable/src/common/exif.cc:3320, function dt_set_xmp_dt_metadata(): prepare "SELECT color FROM main.color_labels WHERE imgid=?1"
138,463996 [sql] /home/nilvus/Projets/darktable/src/common/tags.c:623, function dt_set_darktable_tags(): prepare "SELECT COUNT(*) FROM memory.darktable_tags"
138,464035 [sql] /home/nilvus/Projets/darktable/src/common/tags.c:655, function dt_tag_get_attached(): prepare "SELECT DISTINCT T.id, T.name, T.flags, T.synonyms, 1 AS inb FROM main.tagged_images AS I JOIN data.tags T on T.id = I.tagid WHERE I.imgid = 20700 AND T.id NOT IN memory.darktable_tags ORDER BY T.name"
138,464127 [sql] /home/nilvus/Projets/darktable/src/common/tags.c:1224, function dt_selected_images_count(): prepare "SELECT count(*) FROM main.selected_images"
138,464212 [sql] /home/nilvus/Projets/darktable/src/common/tags.c:623, function dt_set_darktable_tags(): prepare "SELECT COUNT(*) FROM memory.darktable_tags"
138,464235 [sql] /home/nilvus/Projets/darktable/src/common/tags.c:655, function dt_tag_get_attached(): prepare "SELECT DISTINCT T.id, T.name, T.flags, T.synonyms, 1 AS inb FROM main.tagged_images AS I JOIN data.tags T on T.id = I.tagid WHERE I.imgid = 20700 AND T.id NOT IN memory.darktable_tags ORDER BY T.name"
138,464295 [sql] /home/nilvus/Projets/darktable/src/common/tags.c:1224, function dt_selected_images_count(): prepare "SELECT count(*) FROM main.selected_images"
138,464362 [sql] /home/nilvus/Projets/darktable/src/common/exif.cc:3117, function dt_set_xmp_dt_history(): prepare "SELECT imgid, formid, form, name, version, points, points_count, source, num FROM main.masks_history WHERE imgid = ?1 ORDER BY num"
138,464424 [sql] /home/nilvus/Projets/darktable/src/common/exif.cc:3170, function dt_set_xmp_dt_history(): prepare "SELECT module, operation, op_params, enabled, blendop_params,        blendop_version, multi_priority, multi_name, num FROM main.history WHERE imgid = ?1 ORDER BY num"
138,469036 [sql] /home/nilvus/Projets/darktable/src/common/history.c:1392, function dt_history_hash_read(): prepare "SELECT basic_hash, auto_hash, current_hash FROM main.history_hash WHERE imgid = ?1"
138,470026 [sql] /home/nilvus/Projets/darktable/src/common/image.c:2065, function dt_image_write_sidecar_file(): prepare "UPDATE main.images SET write_timestamp = STRFTIME('%s', 'now') WHERE id = ?1"
138,470234 [sql] /home/nilvus/Projets/darktable/src/common/collection.c:1741, function dt_collection_update_query(): prepare "SELECT imgid FROM memory.collected_images WHERE imgid NOT IN (20700) AND rowid>(SELECT rowid FROM memory.collected_images WHERE imgid=20700) ORDER BY rowid LIMIT 1"
138,470362 [sql] /home/nilvus/Projets/darktable/src/common/collection.c:865, function _dt_collection_compute_count(): prepare "SELECT COUNT(DISTINCT mi.id) FROM images AS mi WHERE   (flags & 256) != 256  AND  (flags & 7) >= 0 AND (flags & 7) != 6  AND  (1=1 AND (film_id IN (SELECT id FROM main.film_rolls WHERE folder LIKE '/home/nilvus/Images/Photos/Négatif RAW/2020/04-2020/Seyssinet'))) ORDER BY datetime_taken, group_id , mi.id-group_id != 0, filename, version LIMIT ?1, ?2"
138,470716 [sql] /home/nilvus/Projets/darktable/src/common/collection.c:865, function _dt_collection_compute_count(): prepare "SELECT COUNT(DISTINCT mi.id) FROM images AS mi WHERE   (flags & 256) != 256  AND  (flags & 7) >= 0 AND (flags & 7) != 6  AND  (1=1 AND (film_id IN (SELECT id FROM main.film_rolls WHERE folder LIKE '/home/nilvus/Images/Photos/Négatif RAW/2020/04-2020/Seyssinet'))) ORDER BY datetime_taken, group_id , mi.id-group_id != 0, filename, version LIMIT ?1, ?2"
138,471184 [sql] /home/nilvus/Projets/darktable/src/common/collection.c:914, function dt_collection_get(): prepare "SELECT id FROM main.selected_images AS s JOIN (SELECT DISTINCT mi.id FROM images AS mi WHERE   (flags & 256) != 256  AND  (flags & 7) >= 0 AND (flags & 7) != 6  AND  (1=1 AND (film_id IN (SELECT id FROM main.film_rolls WHERE folder LIKE '/home/nilvus/Images/Photos/Négatif RAW/2020/04-2020/Seyssinet'))) ORDER BY datetime_taken, group_id , mi.id-group_id != 0, filename, version LIMIT ?1, ?2) AS mi WHERE mi.id = s.imgid LIMIT -1, ?3"
138,471955 [sql] /home/nilvus/Projets/darktable/src/common/collection.c:893, function dt_collection_get_selected_count(): prepare "SELECT COUNT(*) FROM main.selected_images"
138,472042 [sql] /home/nilvus/Projets/darktable/src/common/collection.c:1810, function dt_collection_update_query(): prepare "DELETE FROM main.selected_images WHERE imgid NOT IN (SELECT DISTINCT mi.id FROM images AS mi WHERE   (flags & 256) != 256  AND  (flags & 7) >= 0 AND (flags & 7) != 6  AND  (1=1 AND (film_id IN (SELECT id FROM main.film_rolls WHERE folder LIKE '/home/nilvus/Images/Photos/Négatif RAW/2020/04-2020/Seyssinet'))) ORDER BY datetime_taken, group_id , mi.id-group_id != 0, filename, version LIMIT ?1, ?2)"
138,472220 [sql] /home/nilvus/Projets/darktable/src/common/collection.c:160, function dt_collection_memory_update(): exec "DELETE FROM memory.collected_images"
138,472255 [sql] /home/nilvus/Projets/darktable/src/common/collection.c:162, function dt_collection_memory_update(): exec "DELETE FROM memory.sqlite_sequence WHERE name='collected_images'"
138,472299 [sql] /home/nilvus/Projets/darktable/src/common/collection.c:170, function dt_collection_memory_update(): prepare "INSERT INTO memory.collected_images (imgid) SELECT DISTINCT mi.id FROM images AS mi WHERE   (flags & 256) != 256  AND  (flags & 7) >= 0 AND (flags & 7) != 6  AND  (1=1 AND (film_id IN (SELECT id FROM main.film_rolls WHERE folder LIKE '/home/nilvus/Images/Photos/Négatif RAW/2020/04-2020/Seyssinet'))) ORDER BY datetime_taken, group_id , mi.id-group_id != 0, filename, version LIMIT ?1, ?2"
138,472853 [sql] /home/nilvus/Projets/darktable/src/common/collection.c:865, function _dt_collection_compute_count(): prepare "SELECT COUNT(DISTINCT mi.id) FROM images AS mi WHERE   (flags & 256) != 256  AND  (flags & 7) >= 0 AND (flags & 7) != 6  AND  (1=1 AND (film_id IN (SELECT id FROM main.film_rolls WHERE folder LIKE '/home/nilvus/Images/Photos/Négatif RAW/2020/04-2020/Seyssinet'))) ORDER BY datetime_taken, group_id , mi.id-group_id != 0, filename, version"
138,473235 [sql] /home/nilvus/Projets/darktable/src/common/collection.c:865, function _dt_collection_compute_count(): prepare "SELECT COUNT(DISTINCT mi.id) FROM images AS mi WHERE   (flags & 256) != 256  AND  (flags & 7) >= 0 AND (flags & 7) != 6  AND  (1=1 AND (film_id IN (SELECT id FROM main.film_rolls WHERE folder LIKE '/home/nilvus/Images/Photos/Négatif RAW/2020/04-2020/Seyssinet'))) ORDER BY datetime_taken, group_id , mi.id-group_id != 0, filename, version"
138,473711 [sql] /home/nilvus/Projets/darktable/src/common/collection.c:914, function dt_collection_get(): prepare "SELECT id FROM main.selected_images AS s JOIN (SELECT DISTINCT mi.id FROM images AS mi WHERE   (flags & 256) != 256  AND  (flags & 7) >= 0 AND (flags & 7) != 6  AND  (1=1 AND (film_id IN (SELECT id FROM main.film_rolls WHERE folder LIKE '/home/nilvus/Images/Photos/Négatif RAW/2020/04-2020/Seyssinet'))) ORDER BY datetime_taken, group_id , mi.id-group_id != 0, filename, version) AS mi WHERE mi.id = s.imgid LIMIT -1, ?3"
138,474180 [sql] /home/nilvus/Projets/darktable/src/common/collection.c:893, function dt_collection_get_selected_count(): prepare "SELECT COUNT(*) FROM main.selected_images"
138,474260 [sql] /home/nilvus/Projets/darktable/src/dtgtk/thumbtable.c:188, function _thumb_get_rowid(): prepare "SELECT rowid FROM memory.collected_images WHERE imgid=20695"
138,474316 [sql] /home/nilvus/Projets/darktable/src/dtgtk/thumbtable.c:359, function _thumbtable_update_scrollbars(): prepare "SELECT COUNT(*) FROM memory.collected_images"
138,474343 reload thumbs from db. force=1 w=1302 h=923 zoom=4 rows=3 size=325 offset=41 centering=1...
138,474354 [sql] /home/nilvus/Projets/darktable/src/dtgtk/thumbtable.c:1605, function dt_thumbtable_full_redraw(): prepare "SELECT rowid, imgid FROM memory.collected_images WHERE rowid>=41 LIMIT 12"
138,474416 done in 0,0001 sec 0 thumbs reloaded
[mipmap_cache] thumbs fill 3,71/8000,00 MB (0,05%)
[mipmap_cache] float fill 0/4 slots (0,00%)
[mipmap_cache] full  fill 0/4 slots (0,00%)
[mipmap_cache] level | near match | miss | stand-in | fetches | total rq
[mipmap_cache] thumb |  50,00% |  50,00% |   -nan%  |   -nan% | 100,00%
[mipmap_cache] float |   -nan% |   -nan% |   -nan%  |   -nan% |   0,00%
[mipmap_cache] full  |   -nan% |   -nan% |   -nan%  |   -nan% |   0,00%


138,474509 [sql] /home/nilvus/Projets/darktable/src/common/history.c:1476, function dt_history_hash_get_status(): prepare "SELECT CASE  WHEN basic_hash == current_hash THEN 1  WHEN auto_hash == current_hash THEN 2  WHEN (basic_hash IS NULL OR current_hash != basic_hash) AND       (auto_hash IS NULL OR current_hash != auto_hash) THEN 4  ELSE 1 END AS status FROM main.history_hash WHERE imgid = 20700"
138,480901 [sql] /home/nilvus/Projets/darktable/src/libs/collect.c:1577, function list_view(): prepare "SELECT folder, film_rolls_id, COUNT(*) AS count FROM main.images AS mi JOIN (SELECT id AS film_rolls_id, folder       FROM main.film_rolls)   ON film_id = film_rolls_id  WHERE (1=1) GROUP BY folder ORDER BY film_rolls_id DESC"
138,483881 [sql] /home/nilvus/Projets/darktable/src/libs/tools/timeline.c:624, function _time_read_bounds_from_collection(): prepare "SELECT db.datetime_taken FROM main.images AS db, memory.collected_images AS col WHERE db.id=col.imgid AND LENGTH(db.datetime_taken) = 19 AND db.datetime_taken > '0001:01:01 00:00:00' ORDER BY db.datetime_taken ASC LIMIT 1"
138,484113 [sql] /home/nilvus/Projets/darktable/src/libs/tools/timeline.c:645, function _time_read_bounds_from_collection(): prepare "SELECT db.datetime_taken FROM main.images AS db, memory.collected_images AS col WHERE db.id=col.imgid AND LENGTH(db.datetime_taken) = 19 AND db.datetime_taken > '0001:01:01 00:00:00' ORDER BY db.datetime_taken DESC LIMIT 1"
138,484336 [sql] /home/nilvus/Projets/darktable/src/libs/duplicate.c:301, function _lib_duplicate_init_callback(): prepare "SELECT i.version, i.id, m.value FROM images AS i LEFT JOIN meta_data AS m ON m.id = i.id AND m.key = ?3 WHERE film_id = ?1 AND filename = ?2 ORDER BY i.version"
```

**Comment by Nilvus** [comment]
@TurboGit: I've done some git bisect and find that's your second commit of #4779 that cause this issue. reject was working great on first commit of this PR (minor code clean-up commit).

I also add if there's some images rejected and appear on rejected only filter, when this issue is there, images on rejected filter cannot be un-toggle. When clicking (or using shortcut, the reject toggle remains toggled.

See git bisect result:

```
git bisect good ff6440133bb273eb23f9dd6f5f2c3c4ce878ac58
0aefb01570cce5212696e7f970538958b16fdc09 is the first bad commit
commit 0aefb01570cce5212696e7f970538958b16fdc09
Author: Pascal Obry <pascal@obry.net>
Date:   Tue Apr 21 20:21:10 2020 +0200

    Introduce a rejected flag per image and use it.
    
    Having this rejected image stored means that we can keep the actual
    stars of the image before being rejected and we can then recover
    this data when clicking on the reject icon a second time.
    
    This works a bit like ctrl-z (undo) but can also be used across
    session.

 src/common/exif.cc    |  7 +++++--
 src/common/image.h    |  2 ++
 src/common/ratings.c  | 23 ++++++++++++++++-------
 src/common/ratings.h  |  3 +++
 src/dtgtk/thumbnail.c |  2 +-
 src/views/view.c      |  2 +-
 6 files changed, 28 insertions(+), 11 deletions(-)

```

**Comment by TurboGit** [comment]
@Nilvus : Thanks for bisecting this, my commit, my fault :) I'll assign this to me and will fix ASAP.

**Comment by Nilvus** [comment]
@TurboGit: Just an add: I've made all tests with an empty and default darktable and as set of images without any XMP on start.

I now switch back to my config and all my images with the ff6440133bb273eb23f9dd6f5f2c3c4ce878ac58 who reject works great. With this commit all images I had rejected with this issue present are no more rejected. The toggle is no more red and images are not rejected. So I think that this issue only select and activate visually reject toggle. But reject toggle is probably not correctly updated on database (I've check XMP and they have the rating to -1.
